loopchain/blockchain/blockchain.py

```diff
@@ -98,6 +111,60 @@ def __init__(self, channel_name=None, store_id=None, block_manager=None):
         for tx_version, tx_hash_version in channel_option.get("hash_versions", {}).items():
             self.__tx_versioner.hash_generator_versions[tx_version] = tx_hash_version
 
+    @property
+    def leader_made_block_count(self) -> int:
+        if self.__last_block:
+            return self.__made_block_count[self.__last_block.header.peer_id]
+        return -1
+
+    @property
+    def my_made_block_count(self) -> int:
+        return self.__made_block_count[ChannelProperty().peer_address]
+
+    def _up_made_block_count(self, block: Block) -> None:
+        """This is must called before changing self.__last_block!
+
+        :param block:
+        :return:
+        """
+        if block.header.height == 0:
+            return
+
+        if self.__last_block.header.peer_id != block.header.peer_id:
+            self.__made_block_count[block.header.peer_id] = 1
+        else:
+            self.__made_block_count[block.header.peer_id] += 1
+
+        utils.logger.spam(f"({block.header.height})made_block_count:\n{self.__made_block_count}")
+
+    def reset_leader_made_block_count(self):
+        self.__made_block_count.clear()
+
+    def get_next_leader(self) -> str:
+        """get next leader by leader_made_block_count
+
+        :return: new leader's peer_id as hex_hx(str)
+        """
+
+        peer_manager = ObjectManager().channel_service.peer_manager
+        if self.leader_made_block_count == (conf.MAX_MADE_BLOCK_COUNT - 1):
+            # (conf.MAX_MADE_BLOCK_COUNT - 1) means if made_block_count is 9,
+            # next unconfirmed block height is 10 and It has to have changed next leader.
+            return peer_manager.get_next_leader_peer(self.__last_block.header.peer_id.hex_hx()).peer_id
+
+        return peer_manager.leader_id
+
+    def get_expected_generator(self, unconfirmed_block: Block) -> str:
+        """get expected generator to vote unconfirmed block
+
+        :return: expected generator's id by made block count.
+        """
+
+        peer_manager = ObjectManager().channel_service.peer_manager
+        if self.__made_block_count[unconfirmed_block.header.peer_id] > conf.MAX_MADE_BLOCK_COUNT:
+            return peer_manager.get_next_leader_peer(unconfirmed_block.header.peer_id).peer_id
+        return unconfirmed_block.header.peer_id.hex_hx()
```

Comment:
How about pull of `unconfirmed_block.header.peer_id` as variable?
It is repeated 3 times!